### PR TITLE
feat: price from order book snapshot

### LINKED ISSUE #147: Price from order book
listen to order book channel https://www.okx.com/docs-v5/en/#websocket-api-public-channel-order-book-channel

Currently we are transmitting https://github.com/GaloyMoney/stablesats-rs/blob/3b45e2960a1c22a4006aa33f8f1bbb88e96c8498/shared/src/payload/mod.rs#L24
We need a new message to transmit the order book.

Currently we store price in price cache https://github.com/GaloyMoney/stablesats-rs/blob/c6dbbe95b311f19b92a5f9912bb170721e7cb1c4/price-server/src/app/mod.rs#L52.

Now we must cache the whole order book.

Instead of calling https://github.com/GaloyMoney/stablesats-rs/blob/c6dbbe95b311f19b92a5f9912bb170721e7cb1c4/price-server/src/app/mod.rs#L68 `latest_tick` => `latest_snapshot`

If we want to buy $5000 instead of looking at last tick -> we need to run through the order book and figure out the weighted price.

Volume 500 => Price $100
Volume 4000 => Price $101
Volume 500 => Price $102

500 * 100 + 4000 * 101 + 500 * 102 => cost of $5000?

### PR DESCRIPTION
**Pull request by enigbe**
## What this PR does?
- Serves the exchange price from snapshot of the order book

## Currently does (will do) the following
- [x] Streams the order book snapshot and incremental updates
- [x] Creates new message wrapper to transmit snapshots of the order book
- [x] Merges updates to snapshot and publishes to redis

    - [x] Update snapshot with incremental order book data with matching prices
    - [x] Update snapshot with incremental order book data with no matching prices
    - [x] Calculate checksum to validate order book depth data

- [x] Caches the order book 
- [x] Retrieves the latest snapshot and compute weighted price

## Related issues
- Resolves #147

### CODE REVIEW

File: okex-price/src/convert.rs [hunk 1]
```diff
@@ -17,3 +20,55 @@ impl TryFrom<OkexPriceTick> for OkexBtcUsdSwapPricePayload {
         }))
     }
 }
+
+impl From<OrderBookAction> for OrderBookActionRaw {
```
Comment (bodymindarts) [inline okex-price/src/convert.rs:24]: Seems redundant - can we use just 1 type?

File: okex-price/src/lib.rs [hunk 2]
```diff
@@ -1,28 +1,104 @@
...
 }
 
+async fn order_book_subscription(
+    publisher: Publisher,
+    config: &PriceFeedConfig,
+) -> Result<(), PriceFeedError> {
+    let mut stream = subscribe_btc_usd_swap_order_book(config.clone()).await?;
+    let full_load = stream.next().await.ok_or(PriceFeedError::InitialFullLoad)?;
+    let full_incr = OrderBookIncrement::try_from(full_load)?;
+    let cache = OrderBookCache::new(full_incr.try_into()?);
```
Comment (bodymindarts) [inline okex-price/src/lib.rs:57]: `new(NewType::try_into(full_incr))?`

File: okex-price/src/lib.rs [hunk 3]
```diff
@@ -1,28 +1,104 @@
...
+    );
+
+    let (send, recv) = tokio::sync::oneshot::channel();
+
+    shared::tracing::record_error::<(), PriceFeedError, _, _>(
+        tracing::Level::ERROR,
+        || async move {
+            tokio::spawn(async move {
+                while let Some(book) = stream.next().await {
+                    let span = info_span!(
```
Comment (bodymindarts) [inline okex-price/src/lib.rs:73]: Probably too noisy

File: okex-price/src/lib.rs [hunk 4]
```diff
@@ -32,3 +108,20 @@ async fn okex_price_tick_received(
...
 }
+
+async fn okex_order_book_received(
+    publisher: &Publisher,
+    book: OkexOrderBook,
+    mut cache: OrderBookCache,
+) -> Result<(), PriceFeedError> {
+    if let Ok(increment) = OrderBookIncrement::try_from(book) {
+        cache.update_order_book(increment)?;
+        if let Ok(complete_order_book) = OkexBtcUsdSwapOrderBookPayload::try_from(cache.latest()) {
```
Comment (bodymindarts) [inline okex-price/src/lib.rs:119]: Why `try_from`?

File: okex-price/src/lib.rs [hunk 5]
```diff
@@ -32,3 +108,20 @@ async fn okex_price_tick_received(
...
+            publisher
+                .throttle_publish::<OkexBtcUsdSwapOrderBookPayload>(complete_order_book)
```
Comment (bodymindarts) [inline okex-price/src/lib.rs:121]: Is the generic needed?

File: okex-price/src/okex_shared.rs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+#[derive(Clone, Deserialize, Debug, PartialEq, Eq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ChannelArgs {
+    pub channel: String,
+    pub inst_id: String,
+}
+
+#[derive(Clone, Deserialize, Debug, PartialEq, Eq, Serialize)]
+pub struct UnsubscribeResponse {
```
Comment (bodymindarts) [inline okex-price/src/okex_shared.rs:11]: Where is this used?

File: okex-price/src/order_book/book.rs [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+    pub fn new(book: CompleteOrderBook) -> Self {
+        Self { current: book }
+    }
+
+    pub fn update_order_book(&mut self, book: OrderBookIncrement) -> Result<(), PriceFeedError> {
+        self.current = self.current.try_merge(book)?;
+        Ok(())
+    }
+
+    pub fn latest(&self) -> CompleteOrderBook {
```
Comment (bodymindarts) [inline okex-price/src/order_book/book.rs:207]: return `ref`

File: price-server/src/order_book_snapshot_cache.rs [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+///Market Order Book (MOB)
+#[derive(Debug, Clone, Deserialize)]
+pub struct MOBSnapshot {
+    pub asks: BTreeMap<QuotePrice, Decimal>,
+    pub bids: BTreeMap<QuotePrice, Decimal>,
+    pub timestamp: TimeStamp,
+}
+impl From<OrderBookPayload> for MOBSnapshot {
+    fn from(value: OrderBookPayload) -> Self {
+        let mut asks = BTreeMap::new();
```
Comment (bodymindarts) [inline price-server/src/order_book_snapshot_cache.rs:111]:
```
let asks = value.asks.into_iter().map(..).collect()
```

File: price-server/tests/fixtures/order-book-payload-contrived.json [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+        "asks": {
+            "0.0100000": "10"
+        },
+        "bids": {
+            "0.0100000": "10"
+        },
+        "timestamp": 1667454784,
+        "exchange": "okex"
+    }
+}
```
Comment (bodymindarts) [inline price-server/tests/fixtures/order-book-payload-contrived.json:17]: missing newline

File: price-server/src/order_book_snapshot_cache.rs [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+            asks,
+            bids,
+            timestamp: value.timestamp,
+        }
+    }
+}
+
+impl MOBSnapshot {
+    pub fn sell_usd(&self) -> PriceConverter {
+        PriceConverter::new(self.ask_price_of_one_sat())
```
Comment (bodymindarts) [inline price-server/src/order_book_snapshot_cache.rs:136]: `VolumeBasePriceConverter::new(&sel)`

File: price-server/src/order_book_snapshot_cache.rs (hunk 8 above)
Comment (bodymindarts) [reply price-server/src/order_book_snapshot_cache.rs:111]: Must normalize for contract size.

File: price-server/src/order_book_snapshot_cache.rs [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+        Self {
+            stale_after,
+            snapshot: None,
+        }
+    }
+
+    fn update_snapshot(&mut self, snap: Envelope<OkexBtcUsdSwapOrderBookPayload>) {
+        let payload = snap.payload.0;
+
+        if let Some(snap) = self.snapshot.clone() {
```
Comment (bodymindarts) [inline price-server/src/order_book_snapshot_cache.rs:63]: Why clone?

Comment (bodymindarts) [reply price-server/src/order_book_snapshot_cache.rs:63]: `if let Some(snap) = self.snapshot.as_ref()`

File: price-server/src/order_book_snapshot_cache.rs [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+            if snap.timestamp > payload.timestamp {
+                return;
+            }
+        }
+
+        if let Ok(snapshot) = MOBSnapshot::try_from(payload) {
```
Comment (bodymindarts) [inline price-server/src/order_book_snapshot_cache.rs:69]: Should this be `try_from`?

File: price-server/src/order_book_snapshot_cache.rs [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+}
+impl QuotePrice {
+    pub fn inner(&self) -> Decimal {
+        self.0
+    }
+}
+
+///Market Order Book (MOB)
+#[derive(Debug, Clone, Deserialize)]
+pub struct MOBSnapshot {
```
Comment (bodymindarts) [inline price-server/src/order_book_snapshot_cache.rs:104]: `OrderBookView`

File: price-server/src/order_book_snapshot_cache.rs [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+                return;
+            }
+        }
+
+        if let Ok(snapshot) = MOBSnapshot::try_from(payload) {
+            self.snapshot = Some(snapshot);
+        }
+    }
+
+    fn latest_snapshot(&self) -> Result<MOBSnapshot, SnapshotCacheError> {
```
Comment (bodymindarts) [inline price-server/src/order_book_snapshot_cache.rs:74]: `current`

File: price-server/src/order_book_snapshot_cache.rs [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+    #[error("PayloadConversion: conversion from OrderBookPayload failed")]
+    PayloadConversion,
+    #[error("OutdatedSnapshot: last update was at {0}")]
+    OutdatedSnapshot(TimeStamp),
+    #[error("No snapshot data available")]
+    NoSnapshotAvailable,
+}
+
+#[derive(Debug, Clone)]
+pub struct SnapshotCache {
```
Comment (bodymindarts) [inline price-server/src/order_book_snapshot_cache.rs:27]: `OrderBookCache`

**Comment by enigbe** [comment]
Splitting this into two PRs. The first to address streaming and publishing the order books (see [link](https://github.com/GaloyMoney/stablesats-rs/pull/176)) and the second (WIP) to update the  weighted price calculated in this PR

**Pull request closed without merging**